python/python/lance/_dataset/sharded_dataset.py

```diff
@@ -0,0 +1,143 @@
+#  Copyright (c) 2024. Lance Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from pathlib import Path
+from typing import Generator, List, Literal, Optional, Union
+
+import pyarrow as pa
+
+import lance
+from lance.dataset import LanceDataset
+
+__all__ = ["ShardedDataset"]
+
+
+def _sharded_fragments(
+    ds: LanceDataset,
+    batch_size: int,
+    rank: int,
+    world_size: int,
+    columns: Optional[List[str]] = None,
+    batch_readahead: int = 8,
+    with_row_id: bool = False,
+) -> Generator[pa.RecordBatch, None, None]:
+    fragments = ds.get_fragments()
+    for idx in range(rank, len(fragments), world_size):
+        frag = fragments[idx]
+        for batch in frag.to_batches(
+            columns=columns,
+            batch_size=batch_size,
+            with_row_id=with_row_id,
+            batch_readahead=batch_readahead,
+        ):
+            yield batch
+
+
+def _sharded_batches(
+    ds: LanceDataset,
+    batch_size: int,
+    rank: int,
+    world_size: int,
+    columns: Optional[List[str]] = None,
+    with_row_id: bool = False,
+) -> Generator[pa.RecordBatch, None, None]:
+    if with_row_id:
+        raise NotImplementedError("with_row_id is not supported for batch sharding yet")
+
+    total = ds.count_rows()
+
+    def _gen_ranges():
+        for start in range(rank * batch_size, total, world_size * batch_size):
+            yield start, start + batch_size
+
+    return ds._ds.take_scan(
+        _gen_ranges(),
+        columns=columns,
+    )
+
+
+class ShardedDataset:
```

Comment:
`ShardedDataset` sounds like it should be a subclass or ducktype or LanceDataset. Is this just providing an iterator? If so, call it DatasetIterator?